ws-server/src/future.rs

```diff
@@ -130,3 +136,80 @@ where
 		this.selector.poll_unpin(cx)
 	}
 }
+
+#[derive(Debug)]
+struct MonitorInner {
+	shutdown_requested: AtomicBool,
+	waker: AtomicWaker,
+}
+
+/// Monitor for checking whether the server has been flagged to shut down.
+#[derive(Debug, Clone)]
+pub(crate) struct StopMonitor(Arc<MonitorInner>);
+
+impl Drop for StopMonitor {
+	fn drop(&mut self) {
+		if Arc::strong_count(&self.0) == 1 {
+			self.0.waker.wake();
+		}
+	}
+}
+
+impl StopMonitor {
+	pub(crate) fn new() -> Self {
+		StopMonitor(Arc::new(MonitorInner { shutdown_requested: AtomicBool::new(false), waker: AtomicWaker::new() }))
+	}
+
+	pub(crate) fn shutdown_requested(&self) -> bool {
+		self.0.shutdown_requested.load(Ordering::Relaxed)
```

Comment:
The way I read this was it's ok to read the "wrong" value here; it'd simply be polled again on the next iteration.

Reply:
@niklasad1: @dvdplm has the right of it, it's being polled in loops (both for the socket listener and background tasks) so ordering doesn't really matter, and it's better if it's faster since those are relatively hot loops :)

Reply:
ok got it, makes sense :)